### ISSUE TITLE
Closures binding fields broken

### ISSUE THREAD

**Issue by TobiasWrigstad** [parapluu/encore#18]
Code like the following (in active class `PFilter`)

```
let action = \(i:int) -> { this.filter.cancel(i); (); } in 
  primes.for_all(action)
```

generate C code like the following (manually cleaned, hopefully equivalent)

```
pony_actor_t* this = ((struct ____closure0_env*) _env)->this;
Filter_data* _tmp0 = this->filter;
void* _tmp1 = Filter_cancel(_tmp0, i);
```

which is (correctly) rejected because `this` is a `pony_actor_t` which is opaque, meaning `this->filter` is illegal. 

As far as I can tell, the type of `this` should really be `PFilter_data*`in this case. It seems that the implementation captures `this` rather than `this.filter`, and then attempts to extract `this.filter` later.

**Comment by supercooldave** [comment]
As far as I can tell, this issue has been resolved, based on the following test case, which compiles and runs sensibly in the current development branch.

```
local class Filter
  def cancel(i : int) : unit
    println(i)
  end
end

local class Primes
  def forall(f : int -> unit) : unit
    f(10)
  end
end


active class Pif
 
  val filter : Filter

  def init() : unit
    this.filter = new Filter
  end

  def foo() : unit
    val primes = new Primes
    val action = fun (i : int) : unit => this.filter.cancel(i)
    primes.forall(action)
  end
end

active class Main
  def main() : unit
    (new Pif)!foo()
  end
end
```

@TobiasWrigstad @EliasC @kikofernandez Any reason to not mark this as resolved?

**Comment by kikofernandez** [comment]
Is there a test for it in the `tests` directory?

**Comment by supercooldave** [comment]
It seems that there isn't. I will simplify the code above and make a test case.
Before closing, I'll see if anyone else has anything to add.

**Comment by supercooldave** [comment]
Addressed in PR #787.
